From 2c349955b78316d13eee3997fc691bf8827c7917 Mon Sep 17 00:00:00 2001
From: Konstantin Seurer <konstantin.seurer@gmail.com>
Date: Tue, 7 Jan 2025 21:20:42 +0100
Subject: [PATCH 6/9] llvmpipe: Fix half-pixel sample offset with AF

Simply adding -0.5 will cause a noticeable offset for low sample counts.

Reviewed-by: Aleksi Sapon <aleksi.sapon@autodesk.com>
Acked-by: Mike Blumenkrantz <michael.blumenkrantz@gmail.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/32935>

Upstream-Status: Backport [https://gitlab.freedesktop.org/mesa/mesa/-/commit/3f7564d86b1213eb2d54142a73f6e89d72ecd610]
---
 src/gallium/auxiliary/gallivm/lp_bld_sample_soa.c   | 4 +++-
 src/gallium/drivers/llvmpipe/ci/traces-llvmpipe.yml | 6 +++---
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/gallium/auxiliary/gallivm/lp_bld_sample_soa.c b/src/gallium/auxiliary/gallivm/lp_bld_sample_soa.c
index c0a1a140415..50f96544da1 100644
--- a/src/gallium/auxiliary/gallivm/lp_bld_sample_soa.c
+++ b/src/gallium/auxiliary/gallivm/lp_bld_sample_soa.c
@@ -2266,6 +2266,8 @@ lp_build_sample_aniso(struct lp_build_sample_context *bld,
 
    LLVMValueRef float_N = lp_build_int_to_float(coord_bld, N);
    LLVMValueRef rcp_N = lp_build_rcp(coord_bld, float_N);
+   LLVMValueRef base_k = LLVMBuildFMul(builder, float_N, lp_build_const_vec(gallivm, coord_bld->type, -0.5), "");
+   base_k = lp_build_add(coord_bld, base_k, lp_build_const_vec(gallivm, coord_bld->type, 0.5));
 
    struct lp_build_for_loop_state loop_state;
    lp_build_for_loop_begin(&loop_state, gallivm, lp_build_const_int32(gallivm, 0),
@@ -2275,8 +2277,8 @@ lp_build_sample_aniso(struct lp_build_sample_context *bld,
       k = lp_build_broadcast_scalar(int_coord_bld, k);
 
       LLVMValueRef float_k = lp_build_int_to_float(coord_bld, k);
+      float_k = lp_build_add(coord_bld, float_k, base_k);
       float_k = lp_build_mul(coord_bld, float_k, rcp_N);
-      float_k = lp_build_add(coord_bld, float_k, lp_build_const_vec(gallivm, coord_bld->type, -0.5));
 
       LLVMValueRef u_offset = lp_build_mul(coord_bld, float_k, dudk);
       LLVMValueRef v_offset = lp_build_mul(coord_bld, float_k, dvdk);
-- 
2.34.1

